### ISSUE TITLE
feat: evaluate YOLO-based object detection replacement

### ISSUE THREAD

**Issue by sharp231** [sharp231/classifier#3]
### 概要
物体検出モデル（YOLO 等）へのリプレイスを検討する。

### 対象
- 既存モデルとの精度/速度/要件比較
- モデル候補の選定基準を整理

**Comment by sharp231** [comment]
## 概要

現在のCIFAR-10画像分類モデルを物体検出モデルへ置き換える前に、ユーザー価値、データセット、精度、推論速度、運用コスト、UI変更を比較し、移行するかどうかを判断する。

このIssueではYOLOへの移行を前提とせず、画像分類から物体検出へプロダクトを変更する必要があるかを検証する。

## 背景

現在のアプリケーションは、画像全体をCIFAR-10の10クラスのいずれかに分類する。

```text
入力画像
  ↓
画像全体を分類
  ↓
猫、犬、自動車などの候補を表示
```

物体検出モデルでは、画像内の複数物体を位置情報付きで返す。

```text
入力画像
  ↓
複数物体を検出
  ↓
クラス、信頼度、Bounding Boxを表示
```

物体検出は画像分類の単純な高性能版ではなく、入力、出力、データセット、評価指標、UI、APIが変わるプロダクト変更である。

## 最初に確認するプロダクト要件

次の質問へ回答する。

- ユーザーは画像全体の分類だけで困っているか
- 画像内に複数物体があることが主要な問題か
- 物体の位置情報が必要か
- CIFAR-10の10クラスを維持するか
- 対象クラスを拡張するか
- 判定困難を返すだけで問題を解決できないか
- 画像分類モデルの再学習で改善できないか
- Grad-CAMなどの可視化で目的を満たせないか
- 推論時間の増加を許容できるか
- 無料枠のホスティングで運用できるか

具体的なユーザー課題が確認できない場合は、移行しない判断も含める。

## 検討する選択肢

### 案A: 現在の画像分類を維持する

- 現在のCIFAR-10分類を継続する
- confidence calibrationを追加する
- 判定困難を返せるようにする
- TTAを評価する
- モデル自体を改善する

### 案B: 画像分類と物体検出を併存させる

- 通常の画像分類を残す
- 実験機能として物体検出を追加する
- 同じ画像で両方の結果を比較する
- 本番機能へ昇格する条件を定義する

### 案C: 物体検出へ完全移行する

- 画像分類機能を廃止する
- 物体検出向けのAPIとUIへ変更する
- データセットとモデルを変更する
- 既存利用者への影響と移行方法を定義する

最初は案BのPoCとして隔離して検証し、案Cを前提にしない。

## モデル候補の評価基準

YOLOだけに限定せず、要件に応じて候補を比較する。

評価項目:

- 対応クラス
- ライセンス
- 商用・公開利用条件
- モデルサイズ
- 推論時間
- CPU推論性能
- メモリ使用量
- 導入依存パッケージ
- Pythonバージョンとの互換性
- TensorFlow/Kerasとの共存
- モデル配布方法
- 保守状況
- 脆弱性情報
- エクスポート形式
- ホスティング制約

モデルの名称やバージョンだけで決めず、公式情報と実測結果を記録する。

## データセット

物体検出には、クラスラベルだけでなくBounding Box付きの教師データが必要になる。

次を確認する。

- 使用するデータセット
- 対応クラス
- データセットのライセンス
- 学習・評価への利用条件
- 学習データと評価データの分離
- クラスごとのデータ数
- 小さい物体への対応
- 複数物体画像への対応
- 重複画像
- アノテーション品質
- データセットのバージョン

CIFAR-10の分類精度と物体検出のmAPは直接比較できないため、評価指標を分離する。

## PoCの実装範囲

PoCは既存アプリケーションから分離する。

候補構成:

```text
experiments/
└── object_detection/
    ├── README.md
    ├── evaluate.py
    ├── sample_images/
    └── reports/
```

PoCで確認する。

- モデルを読み込める
- サンプル画像を推論できる
- Bounding Boxを取得できる
- 複数物体を検出できる
- CPU環境での推論時間
- メモリ使用量
- モデル読み込み時間
- 出力形式
- 代表的な失敗例

PoCの時点では既存の `/result` を変更しない。

## API候補

物体検出を実装する場合の出力候補:

```json
{
  "model_version": "detector-v1",
  "inference_ms": 120,
  "image": {
    "width": 640,
    "height": 480
  },
  "detections": [
    {
      "label": "dog",
      "confidence_score": 0.87,
      "bounding_box": {
        "x": 120,
        "y": 80,
        "width": 240,
        "height": 300
      }
    }
  ]
}
```

座標の基準、画像リサイズとの関係、0〜1の正規化座標を使うかを明確にする。

## フロントエンドへの影響

物体検出を採用する場合、次のUIが必要になる。

- Bounding Boxの描画
- 複数物体の一覧
- 各物体のクラスと信頼度
- 物体が検出されなかった状態
- 重なったBounding Boxの表示
- 小さい画面での表示
- 元画像と前処理後画像の座標変換
- アクセシビリティを考慮したテキスト結果

Canvas、SVG、画像への直接描画のどれを使用するかを比較する。

## 評価指標

物体検出モデルでは、最低限次を記録する。

- mAP
- precision
- recall
- クラス別AP
- 検出漏れ
- 誤検出
- 推論時間
- モデル読み込み時間
- メモリ使用量
- モデルファイルサイズ
- CPU環境での実行可否
- ホスティング費用への影響

現在の分類モデルについては、別に次を記録する。

- Top-1 Accuracy
- Top-3 Accuracy
- Log Loss
- クラス別Accuracy
- 推論時間

異なる指標を1つの数値で無理に比較しない。

## セキュリティ

- 既存の画像形式・サイズ・寸法検証を維持する
- モデルファイルの取得元を固定する
- 信頼できないモデルファイルを読み込まない
- モデルファイルのハッシュまたはバージョンを記録する
- 推論時間とメモリに上限を設ける
- 過大画像をモデルへ直接渡さない
- Bounding Box出力をHTMLへ直接連結しない
- 外部モデル取得時に秘密情報をログへ出力しない

## Go/No-Go判断

### Go条件

- 物体の位置情報を必要とするユーザー課題が確認できる
- 分類モデルでは解決できない
- 対象クラスを定義できる
- 利用可能なデータセットがある
- ライセンス条件を満たす
- CPUまたは利用可能な環境で許容時間内に推論できる
- 無料枠または許容費用内で運用できる
- UIとAPIの変更量を許容できる

### No-Go条件

- 位置情報を必要とするユースケースがない
- 「判定困難」の追加で主要課題を解決できる
- 推論時間またはメモリが許容範囲を超える
- 必要なデータセットを用意できない
- ライセンス条件を満たせない
- 無料枠で安定運用できない
- 既存機能への影響が大きすぎる

## スコープ外

このIssue内では次を行わない。

- 本番モデルの置き換え
- 既存分類機能の削除
- 本番APIの変更
- 本番UIへのBounding Box追加
- 新しいモデルの自動デプロイ
- ユーザー画像の収集
- 独自データセットの大規模作成
- GPU環境の契約
- 本番依存パッケージの無承認追加

## 完了条件

- 物体検出を必要とするユーザー課題が文章化されている
- 画像分類との違いが整理されている
- モデル候補の選定基準が定義されている
- データセットとライセンスを確認している
- PoCが既存アプリケーションから分離されている
- 精度・速度・メモリ・モデルサイズを測定している
- フロントエンドとAPIへの影響を整理している
- Go/No-Go判断と根拠が記録されている
- No-Goの場合もIssueを完了できる
- 本番コードを変更せず検証を完了している

## 子Issue候補

- `docs: 物体検出のユースケースを定義する`
- `docs: 物体検出モデルとデータセットを比較する`
- `test: 物体検出モデルのPoCを作成する`
- `perf: 物体検出モデルのCPU性能を評価する`
- `docs: Bounding Box APIの形式を定義する`
- `feat: 物体検出結果の表示方法を検証する`

## 依存関係

- 通常推論とTTAの精度・速度比較
- 信頼度の校正と判定困難表示

既存分類モデルの改善余地を確認した後に判断する。
